osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,370 @@
+
+  // Size is given in bytes.
+  const auto size = prefetch_file_info->FileNameStringsSize;
+  const auto offset = prefetch_file_info->FileNameStringsOffset;
+  if (offset > data.size()) {
+    // Unexpected offset.
+    return result;
+  }
+
+  size_t total_length{0};
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data[0] + offset);
+  while (*next != '\0') {
+    auto length = wcsnlen_s(next, (size - total_length) / 2);
+    if (length == 0 || length == (size - total_length) / 2) {
+      // A null wide character was not found.
+      break;
+    }
+
+    auto filename = wstringToString(next);
+    filenames.emplace_back(std::move(filename));
+    total_length += (length + 1) * 2;
+    if (total_length >= size) {
+      break;
+    }
+    next += length + 1;
+  }
+
+  result.files = osquery::join(filenames, ",");
+  result.files_count = filenames.size();
+  return result;
+}
+
+PrefetchVolumeInfo parseVolumeInfo(
+    const std::vector<UCHAR>& data,
+    const PREFETCH_FILE_INFORMATION* prefetch_file_info,
+    DWORD version) {
+  PrefetchVolumeInfo result;
+
+  const auto volume_header_size = (version == 30) ? 104 : 96;
+  const auto volume_offset = prefetch_file_info->VolumeInformationOffset;
+  // Size is given in bytes.
+  const auto volume_size = prefetch_file_info->VolumesInformationSize;
+  if (volume_offset > data.size()) {
+    // Unexpected offset.
+    return result;
+  }
+
+  // Headers are stacked sequentially.
+  auto volume_header_offset = volume_offset;
+
+  std::vector<std::string> directories;
+  std::vector<std::string> volume_serials;
+  std::vector<std::string> volume_creations;
+  for (size_t i = 0; i < prefetch_file_info->NumberOfVolumes; i++) {
+    if (volume_header_offset + volume_header_size > data.size()) {
+      // Unexpected size.
+      return result;
+    }
+
+    const auto prefetch_volume_info =
+        (PPREFETCH_VOLUME_INFORMATION)(&data[0] + volume_header_offset);
+    const auto creation =
+        filetimeToUnixtime(prefetch_volume_info->VolumeCreationTime);
+    volume_creations.push_back(std::to_string(creation));
+    const auto serial = prefetch_volume_info->VolumeSerialNumber;
+    volume_serials.push_back((boost::format("%x") % serial).str());
+    volume_header_offset += volume_header_size;
+
+    const auto dir_count = prefetch_volume_info->NumberOfDirectoryStrings;
+    size_t dir_offset = prefetch_volume_info->DirectoryStringsOffset;
+    if (dir_count == 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < dir_count; j++) {
+      if (volume_offset + dir_offset + sizeof(PDIRECTORY_STRING) >
+          data.size()) {
+        // Unexpected offset.
+        break;
+      }
+
+      const auto prefetch_directory =
+          (PDIRECTORY_STRING)(&data[0] + volume_offset + dir_offset);
+      dir_offset += sizeof(DIRECTORY_STRING);
+
+      auto length = wcsnlen_s(prefetch_directory->Directory,
+                              (volume_size - dir_offset) / 2);
```

Comment:
instead of `/2` let's use `/sizeof(WCHAR)`

Reply:
switched to sizeof